### PR TITLE
add support for searching TiKV slow logs

### PR DESCRIPTION
**Pull request by Fullstop000**
UCP https://github.com/pingcap-incubator/tidb-dashboard/issues/303

Signed-off-by: Fullstop000 <fullstop1005@gmail.com>

### CODE REVIEW

File: pkg/apiserver/logsearch/pack.go [hunk 1]
```diff
@@ -76,24 +78,28 @@ func serveTaskForDownload(task *TaskModel, c *gin.Context) {
...
+			if err = dumpLog(*task.SlowLogStorePath, tw); err != nil {
+				_ = c.Error(err)
+				return
+			}
+		}
+	}()
+	contentType := "application/tar"
 	extraHeaders := map[string]string{
-		"Content-Disposition": fmt.Sprintf(`attachment; filename="%s"`, stat.Name()),
+		"Content-Disposition": `attachment; filename="logs.tar"`,
```
Comment (breeswish) [inline pkg/apiserver/logsearch/pack.go:100]: I think the previous implementation is better, which always generates a unique name each download

File: pkg/apiserver/logsearch/task.go [hunk 2]
```diff
@@ -73,8 +74,8 @@ func (tg *TaskGroup) SyncRun() {
 	dir := path.Join(tg.service.logStoreDirectory, strconv.Itoa(int(tg.model.ID)))
 	if err := os.MkdirAll(dir, 0777); err == nil {
 		tg.model.LogStoreDir = &dir
-		tg.service.db.Save(tg.model)
 	}
+	tg.service.db.Save(tg.model)
```
Comment (breeswish) [inline pkg/apiserver/logsearch/task.go:78]: If Mkdir failed, previous implementation will stop the later operations

Comment (Fullstop000) [reply pkg/apiserver/logsearch/task.go:78]: Just forgot revert this..

File: pkg/apiserver/logsearch/pack.go [hunk 3]
```diff
@@ -87,13 +90,28 @@ func serveTaskForDownload(task *TaskModel, c *gin.Context) {
...
-
-	contentType := "application/zip"
+	reader, writer := io.Pipe()
+	go func() {
+		tw := tar.NewWriter(writer)
+		defer writer.Close()
+		defer tw.Close()
+		err := dumpLog(*task.LogStorePath, tw)
+		if err != nil {
+			_ = c.Error(err)
```
Comment (Deardrops) [inline pkg/apiserver/logsearch/pack.go:100]: Log error here, do not send the error message to client.

File: pkg/apiserver/logsearch/task.go [hunk 4]
```diff
@@ -211,8 +231,7 @@ func (t *Task) SyncRun() {
 		}
 		for _, msg := range res.Messages {
 			line := logMessageToString(msg)
-			// TODO: use unsafe here: string -> []byte
-			_, err := bufWriter.Write([]byte(line))
+			_, err := bufWriter.Write(*(*[]byte)(unsafe.Pointer(&line)))
```
Comment (Deardrops) [inline pkg/apiserver/logsearch/task.go:234]:
```
*(*[]byte)(unsafe.Pointer(&line))
```
wrapper it to a common function

Comment (Fullstop000) [reply pkg/apiserver/logsearch/task.go:234]: I think just using inline converting is more clear semantically

**Comment by Fullstop000** [comment]
@Deardrops @breeswish PTAL

**Pull request merged**